### PR TITLE
Disable toolchain discovery and provisioning by default for all tests

### PR DESCRIPTION
**Pull request by bmuskalla**
This is to avoid flakiness in tests that may pick up multiple JVMs from the same version but expect a specific one (e.g. `TestTaskToolchainIntegrationTest`)

### CODE REVIEW

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/AbstractIntegrationSpec.groovy [hunk 1]
```diff
@@ -95,10 +95,12 @@ class AbstractIntegrationSpec extends Specification {
     def setup() {
         m2.isolateMavenLocalRepo(executer)
         executer.beforeExecute {
-            executer.withArgument("-Dorg.gradle.internal.repository.max.tentatives=$maxHttpRetries")
+            withArgument("-Dorg.gradle.internal.repository.max.tentatives=$maxHttpRetries")
             if (maxUploadAttempts != null) {
-                executer.withArgument("-Dorg.gradle.internal.network.retry.max.attempts=$maxUploadAttempts")
+                withArgument("-Dorg.gradle.internal.network.retry.max.attempts=$maxUploadAttempts")
             }
+            withArgument("-Porg.gradle.java.installations.auto-detect=false")
```
Comment (wolfs) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/AbstractIntegrationSpec.groovy:102]: I think it would be better to have a boolean in `AbstractGradleExecuter` which disables toolchain auto detect and auto download. It could be something like `disableToolchainDownloadAndDetection` which would be `true` by default. When `true`, `getAllArgs()` would add the two flags as gradle properties (with `false`), when `false`, `getAllArgs()` would add no flags.

This way, tests which test the toolchain detection or auto download can pass the proper flags to the daemon, while all other tests disable detection and auto-download. It may make sense to introduce two boolean flags on the executor as well, if it makes to set the detection and auto download flag different in tests.

The important part is here that there is way to not pass any properties with regards to detection/auto-download to the daemon and the test can the explicitly set what it wants to test.

File: subprojects/platform-jvm/src/integTest/groovy/org/gradle/jvm/toolchain/JavaToolchainDownloadIntegrationTest.groovy [hunk 2]
```diff
@@ -97,7 +95,6 @@ class JavaToolchainDownloadIntegrationTest extends AbstractIntegrationSpec {
         """
 
         propertiesFile << """
-            org.gradle.java.installations.auto-detect=false
```
Comment (wolfs) [inline subprojects/platform-jvm/src/integTest/groovy/org/gradle/jvm/toolchain/JavaToolchainDownloadIntegrationTest.groovy:100]: So this would be a test where we would not disable auto-download by default, but leave detection disabled, right?

**Review by wolfs — CHANGES_REQUESTED** [review]

File: subprojects/platform-jvm/src/integTest/groovy/org/gradle/jvm/toolchain/JavaToolchainDownloadIntegrationTest.groovy (hunk 2 above)
Comment (bmuskalla) [reply subprojects/platform-jvm/src/integTest/groovy/org/gradle/jvm/toolchain/JavaToolchainDownloadIntegrationTest.groovy:100]: Exactly

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/AbstractIntegrationSpec.groovy (hunk 1 above)
Comment (bmuskalla) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/AbstractIntegrationSpec.groovy]: Good call. Was convinced at first but it looks a lot better now with the fluent API on the executor.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 3]
```diff
@@ -818,6 +830,19 @@ public GradleExecuter withWelcomeMessageEnabled() {
         return this;
     }
 
+    @Override
+    public GradleExecuter withToolchainDetectedEnabled() {
```
Comment (wolfs) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java:834]:
```suggestion
    public GradleExecuter withToolchainDetectionEnabled() {
```

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 4]
```diff
@@ -818,6 +830,19 @@ public GradleExecuter withWelcomeMessageEnabled() {
...
+        disableToolchainDetection = false;
+        return this;
+    }
+
+    @Override
+    public GradleExecuter withToolchainDownloadEnabled() {
+        withToolchainDetectedEnabled();
```
Comment (wolfs) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java:841]: Why does enabling toolchain download also enable detection?

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 5]
```diff
@@ -997,6 +1022,13 @@ private boolean noDaemonArgumentGiven() {
             allArgs.add("--warning-mode=" + warningMode.toString().toLowerCase(Locale.ENGLISH));
         }
 
+        if(disableToolchainDownload) {
```
Comment (wolfs) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java:1025]:
```suggestion
        if (disableToolchainDownload) {
```

Nit

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 6]
```diff
@@ -997,6 +1022,13 @@ private boolean noDaemonArgumentGiven() {
...
+            allArgs.add("-Porg.gradle.java.installations.auto-download=false");
+        }
+        if(disableToolchainDetection) {
```
Comment (wolfs) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java:1028]:
```suggestion
        if (disableToolchainDetection) {
```
Nit

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 7]
```diff
@@ -417,6 +422,13 @@ public GradleExecuter copyTo(GradleExecuter executer) {
             executer.withWelcomeMessageEnabled();
         }
 
+        if(!disableToolchainDetection) {
```
Comment (wolfs) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java:425]:
```suggestion
        if (!disableToolchainDetection) {
```
Nit

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 8]
```diff
@@ -417,6 +422,13 @@ public GradleExecuter copyTo(GradleExecuter executer) {
...
+            executer.withToolchainDetectedEnabled();
+        }
+        if(!disableToolchainDownload) {
```
Comment (wolfs) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java:428]:
```suggestion
        if (!disableToolchainDownload) {
```
Nit

**Review by wolfs — COMMENTED** [review]
This looks good! I left some comments, mostly naming/formatting issues.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java (hunk 4 above)
Comment (bmuskalla) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java:841]: Because we need the detection mechanism to detect the newly provisioned toolchain

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java (hunk 5 above)
Comment (bmuskalla) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java:1025]: We really need either IJ auto formatter or spotless :)

File: subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiClientJdkCompatibilityTest.groovy [hunk 9]
```diff
@@ -234,7 +237,8 @@ public class ToolingApiCompatibilityBuildAction implements BuildAction<String> {
         }
 
         when:
-        succeeds("buildAction",
+        succeeds("" +
+            "buildAction",
```
Comment (wolfs) [inline subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiClientJdkCompatibilityTest.groovy:241]: Is this change intentional? I.e. splitting the string in `"" + "buildAction"`?

**Review by wolfs — APPROVED** [review]
LGTM!

File: subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiClientJdkCompatibilityTest.groovy (hunk 9 above)
Comment (bmuskalla) [reply subprojects/tooling-api/src/integTest/groovy/org/gradle/integtests/tooling/ToolingApiClientJdkCompatibilityTest.groovy:241]: Thanks for catching! This was not intentional
